### PR TITLE
Add support for MacroCat Keyboard

### PR DESCRIPTION
**Pull request by starcatmeow**
## Description

Add support for [MacroCat Keyboard](https://github.com/catmunch/macrocat)

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/macrocat/config.h [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+/* rotary encoder */
+#define ENCODERS_PAD_A { D3 }
+#define ENCODERS_PAD_B { D2 }
+#define ENCODER_SWITCH B7
+
+#define VENDOR_ID 0x2022
+#define PRODUCT_ID 0x8086
+#define MANUFACTURER    CatMunch
+#define PRODUCT         MacroCat Keyboard
```
Comment (zvecr) [inline keyboards/macrocat/config.h:16]:
```suggestion
```

File: keyboards/macrocat/keymaps/default/keymap.c [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        {KC_NO,     KC_NO,      KC_NO,  KC_NO},
+        {KC_NO,     KC_NO,      KC_NO,  KC_NO},
+        {KC_NO,     KC_NO,      KC_NO,  KC_NO}
+    },
+    [3] = {
+        {KC_NO,     KC_NO,      KC_NO,  KC_NO},
+        {KC_NO,     KC_NO,      KC_NO,  KC_NO},
+        {KC_NO,     KC_NO,      KC_NO,  KC_NO},
+        {KC_NO,     KC_NO,      KC_NO,  KC_NO}
+    }
```
Comment (zvecr) [inline keyboards/macrocat/keymaps/default/keymap.c:30]:
```suggestion
    }
```

File: keyboards/macrocat/macrocat.h [hunk 3]
```diff
@@ -0,0 +1,6 @@
+// Copyright 2022 catmunch (@catmunch)
```
Comment (zvecr) [inline keyboards/macrocat/macrocat.h:1]: This file should be deleted

File: keyboards/macrocat/rules.mk [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+# Bootloader selection
+BOOTLOADER = qmk-dfu
+
+# Build Options
+#   change yes to no to disable
+#
+ENCODER_ENABLE = yes
+VIA_ENABLE = yes
+RAW_ENABLE = yes
+DYNAMIC_KEYMAP_ENABLE = yes
```
Comment (zvecr) [inline keyboards/macrocat/rules.mk:13]:
```suggestion
# This file intentionally left blank
```

File: keyboards/macrocat/info.json [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    "maintainer": "Catmunch",
+    "bootloader": "atmel-dfu",
+    "diode_direction": "COL2ROW",
+    "features": {
+        "bootmagic": true,
+        "command": false,
+        "console": false,
+        "extrakey": true,
+        "mousekey": true,
+        "nkro": true
```
Comment (zvecr) [inline keyboards/macrocat/info.json:13]:
```suggestion
{
    "manufacturer": "Catmunch",
    "keyboard_name": "MacroCat Keyboard",
    "maintainer": "Catmunch",
    "bootloader": "atmel-dfu",
    "diode_direction": "COL2ROW",
    "features": {
        "bootmagic": true,
        "command": false,
        "console": false,
        "extrakey": true,
        "mousekey": true,
        "nkro": true,
        "encoder": true
```

**Review by zvecr — CHANGES_REQUESTED** [review]

File: keyboards/macrocat/keymaps/default/keymap.c (hunk 2 above)
Comment (starcatmeow) [reply keyboards/macrocat/keymaps/default/keymap.c:30]: Thanks for your review, but in our design these two layers are for letting users put their own macros. Deleting these will cause some storage problems of layer 2&3 in VIA.

Comment (zvecr) [reply keyboards/macrocat/keymaps/default/keymap.c:30]: This is not a concern of a default keymap, there should be a via keymap.

File: keyboards/macrocat/keymaps/default/rules.mk [hunk 6]
```diff
@@ -0,0 +1,3 @@
+VIA_ENABLE = yes
```
Comment (zvecr) [inline keyboards/macrocat/keymaps/default/rules.mk]: This file should be removed

File: keyboards/macrocat/keymaps/oled/rules.mk [hunk 7]
```diff
@@ -0,0 +1,5 @@
+VIA_ENABLE = yes
+RAW_ENABLE = yes
+DYNAMIC_KEYMAP_ENABLE = yes
```
Comment (zvecr) [inline keyboards/macrocat/keymaps/oled/rules.mk:3]:
```suggestion
```

File: keyboards/macrocat/macrocat.c [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    tap_code(KC_MPLY);
+}
+void encoder_double_click(void) {
+    tap_code(KC_MNXT);
+}
+void encoder_triple_click(void) {
+    tap_code(KC_MPRV);
+}
+void matrix_init_kb() {
+    setPinInputHigh(ENCODER_SWITCH);
```
Comment (zvecr) [inline keyboards/macrocat/macrocat.c:126]:
```suggestion
    matrix_init_user();

    setPinInputHigh(ENCODER_SWITCH);
```

File: keyboards/macrocat/keymaps/default/keymap.c (hunk 2 above)
Comment (starcatmeow) [reply keyboards/macrocat/keymaps/default/keymap.c]: Okay, I created a separate via keymap.

File: keyboards/macrocat/macrocat.c [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+          0,  0,  0,  0,  0,  0,  0,  0,  0,  1,  1,  1,  1, 97, 97,121,121,135,135,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+          0,  0,  0,  0,  0,  0,  0,  0,  0,  6,  6, 24, 24, 24, 24, 24, 24,  7,  7,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+    }
+};
+static uint8_t key_pressed_l = 0;
+static uint8_t key_pressed_r = 0;
+oled_rotation_t oled_init_kb(oled_rotation_t rotation) {
+    return OLED_ROTATION_180;
+}
+bool oled_task_kb(void) {
```
Comment (zvecr) [inline keyboards/macrocat/macrocat.c:72]:
```suggestion
bool oled_task_kb(void) {
    if (!oled_task_user()) {
        return false;
    }

```

**Review by zvecr — CHANGES_REQUESTED** [review]

File: keyboards/macrocat/keymaps/via/rules.mk [hunk 10]
```diff
@@ -0,0 +1,3 @@
+VIA_ENABLE = yes
+RAW_ENABLE = yes
+DYNAMIC_KEYMAP_ENABLE = yes
```
Comment (zvecr) [inline keyboards/macrocat/keymaps/via/rules.mk:3]:
```suggestion
```

**Review by zvecr — CHANGES_REQUESTED** [review]
`keyboards/macrocat/keymaps/default/rules.mk` should be deleted and not an empty file

**Comment by starcatmeow** [comment]
> `keyboards/macrocat/keymaps/default/rules.mk` should be deleted and not an empty file

Thanks. Deleted.

**Comment by starcatmeow** [comment]
> `keyboards/macrocat/keymaps/default/rules.mk` should be deleted and not an empty file

Hi @zvecr , is there anything missing now?

**Comment by starcatmeow** [comment]
Hi @zvecr, I can't find how to mark the change requested as resolved, but I've already deleted that file. Can this PR get merged?

**Comment by zvecr** [comment]
> Pull Request reviews are something that is done in our free time. We are not paid nor compensated for the time we spend reviewing, as it is a labor of love. As such, this means that it can take time for us to get to your Pull Request. Things like family, or life can get in the way of us getting to PRs, and burnout is a serious consideration. The QMK firmware repository averages 200 PRs opened and 200 PRs merged every month, so please have patience.

PRs need 2 approvals before they can be merged. Opening and closing new PRs is only going to make the process take longer.

**Review by drashna — APPROVED** [review]

File: keyboards/macrocat/readme.md [hunk 11]
```diff
@@ -0,0 +1,27 @@
+# macrocat
+
+![macrocat](https://i.imgur.com/ItLHGBf.jpg)
```
Comment (drashna) [inline keyboards/macrocat/readme.md:3]:
```suggestion
![macrocat](https://i.imgur.com/ItLHGBfh.jpg)
```

Comment (starcatmeow) [reply keyboards/macrocat/readme.md:3]: Thanks for your advice, but what does the extra h do?

File: keyboards/macrocat/keymaps/default/keymap.c [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+        {KC_NO,     KC_NO,      KC_NO,  KC_NO},
+        {KC_NO,     KC_NO,      KC_NO,  KC_NO},
+        {KC_NO,     KC_NO,      KC_NO,  KC_NO},
+        {KC_NO,     KC_NO,      KC_NO,  KC_NO}
+    },
+    [3] = {
+        {KC_NO,     KC_NO,      KC_NO,  KC_NO},
+        {KC_NO,     KC_NO,      KC_NO,  KC_NO},
+        {KC_NO,     KC_NO,      KC_NO,  KC_NO},
+        {KC_NO,     KC_NO,      KC_NO,  KC_NO}
```
Comment (zvecr) [inline keyboards/macrocat/keymaps/default/keymap.c:29]:
```suggestion
```

**Review by zvecr — APPROVED** [review]

**Pull request merged**